Core/Quest: Only delete bound quest items on turn in if the following

runningnak3d opened the pull request:
**Changes proposed:**
-  Delete ALL bound quest items on turn in only if they meet the following conditions, otherwise delete only the required amount for the quest:
1. Not repeatable 
2. The item isn't a requirement of another quest in the players quest log  

**Target branch(es):** 3.3.5/6.x

**Issues addressed:** Closes #17758 #17806 

**Tests performed:** (Does it build, tested in-game, etc)
- Builds
- Tested with quests 4730 and 4733
  - .quest add 4730
  - .quest add 4733
  - Turn in 4730, and the second bones are still in your bag
- Tested with quest 582 after modifying the quest requirement to be item 12242 (seemed fitting)
- Test 1:
  - .quest add 582
  - .add 12242 21
  - Turn in quest and all 21 bones were deleted
- Test 2 (this is why I changed the requirement to item 12242 -- no other quest needs the heads):
  - .quest add 582
  - .quest add 4730
  - .quest add 4733
  - .add 12242 26
  - Turn in quest and only the required amount were deleted, leaving bones to be turned in for the other quests.

**Known issues and TODO list:**
- [x]  Check if it works
- [x] Test additional quests to make sure that nothing was broken. 
- I removed the fringe case WIP because 1 - I don't think there are any. 2 - if there are, someone will have to open a ticket if they find them

runningnak3d committed:
Core/Quest: Only delete bound quest items on turn in if the following
requirements are met:
* Quest is not repeatable
* The item isn't usable by another quest in the players quest log

jackpoz commented:
> Have no ability to test this in game since I am at work :)

please find the time to test it

runningnak3d committed:
Core/Quest: Logic fail / Copy paste fail

runningnak3d committed:
Core/Quest: C/P fail part 2

runningnak3d committed:
Core/Quest: Hopefully the last fix for quest turnins. A follow up PR
will be done to fix quest abandon / fail

runningnak3d commented:
I have tested every quest that was listed in the original issues that brought on the original PR, and I have tested all the quests listed as issues since then. I need other people to test to make sure there are no fringe cases.

runningnak3d committed:
Fail removing old code

runningnak3d committed:
Spaceing

runningnak3d committed:
Core/Quest: Use default value so original functions don't need to be
changed (Thanks @areiel for the suggestion).

runningnak3d committed:
*sigh* should really compile before pushing so stupid crap like this is
caught...

runningnak3d committed:
Effing typo

jackpoz commented:
do the last changes require a new session of tests ? since it's the last version that is going to be merged

runningnak3d commented:
The last version is the one that was tested -- and passed :)

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -16440,12 +16441,16 @@ bool Player::HasQuestForItem(uint32 itemid) const
 
+            if (questid == excludeQuestId)
+                excludeQuestId = 0;
+                return true;
```
passed my ass.

runningnak3d replied:
Yes, it was tested with my cock in your ass. What the fuck is wrong with you dude?

DDuarte replied:
Brackets are needed. It's currently always returning true. Indentation alone is not enough.

runningnak3d replied:
Yea, I get that. It just so happens that always returning true just happened to make it look like the patch was working as intended. As I explained to Shauren in IRC. A simple, "Hey, some braces here...." would have been enough. If I say I test something, then I test it.

jackpoz replied:
Please debug your code line by line and add breakpoints to ensure all cases are tested.
There's no need to say bad words to who is reviewing your code for mistakes you made.

runningnak3d committed:
This version was NOT tested -- unlike the last one. However, I am
pretty freaking sure that it will work just fine -- JUST LIKE THE LAST
VERSION!

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -16440,12 +16441,18 @@ bool Player::HasQuestForItem(uint32 itemid) const
         QuestStatusData const& q_status = qs_itr->second;
 
-        if (q_status.Status == QUEST_STATUS_INCOMPLETE)
+        if (q_status.Status == QUEST_STATUS_INCOMPLETE || q_status.Status == QUEST_STATUS_COMPLETE)
```
Including QUEST_STATUS_COMPLETE here allows quest items to drop even if player completes the objective, see the use of Player::HasQuestForItem in LootItem::AllowedForPlayer

runningnak3d replied:
Now that was a constructive comment. That is very easily fixed by adding a bool because I need to check for both, but yes, I see how that breaks other uses.

Aokromes commented:
What's missing here?

runningnak3d committed:
Core/Quest: Final solution -- I think. I am not sure I am entirely happy
with it.

runningnak3d committed:
Core/Quest: Remove commented fail attempt

tkrokli commented:
@runningnak3d : you need to rebase your repo on current 3.3.5 source to solve the conflicts. There has been a lot of changes in the core files since you opened this PR, so you need to incorporate the changes into your PR branch files.

runningnak3d committed:
Merge remote-tracking branch 'upstream/3.3.5' into quest6

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -16445,7 +16445,7 @@ bool Player::HasQuestForItem(uint32 itemid) const
         QuestStatusData const& q_status = qs_itr->second;
 
-        if (q_status.Status == QUEST_STATUS_INCOMPLETE)
+        if ((q_status.Status == QUEST_STATUS_INCOMPLETE) || (turnIn && (q_status.Status == QUEST_STATUS_INCOMPLETE || q_status.Status == QUEST_STATUS_COMPLETE)))
```
You don't need to check `q_status.Status == QUEST_STATUS_INCOMPLETE` twice

runningnak3d replied:
Thank you

runningnak3d committed:
Trailing space

runningnak3d committed:
Core/Quest: Remove redundant check -- Thanks Shauren

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -16460,7 +16460,7 @@ bool Player::HasQuestForItem(uint32 itemid) const
             for (uint8 j = 0; j < QUEST_ITEM_OBJECTIVES_COUNT; ++j)
             {
-                if (itemid == qinfo->RequiredItemId[j] && q_status.ItemCount[j] < qinfo->RequiredItemCount[j])
+                if ((itemid == qinfo->RequiredItemId[j] && q_status.ItemCount[j] < qinfo->RequiredItemCount[j]) || (turnIn && itemid == qinfo->RequiredItemId[j] && q_status.ItemCount[j] <= qinfo->RequiredItemCount[j]))
```
The same applies here, duplicated `itemid == qinfo->RequiredItemId[j]` is not needed
`if (itemid == qinfo->RequiredItemId[j] && (q_status.ItemCount[j] < qinfo->RequiredItemCount[j] || (turnIn && q_status.ItemCount[j] <= qinfo->RequiredItemCount[j])))`
I would also want to know why are you using `<=` comparison during quest turn in, what does it change?

runningnak3d replied:
Yea, I saw this after I already committed the removal of the other redundant check, but I had already logged out for the day. As soon as I can get back on my dev box I was going to push it. But again, thanks.

The <= is needed because this check would fail if (for example) you had 2 quests that needed item 1234 and you had 4 of that item. The first quest needs 2, and the second quest needs 2. If you just check for < it would fail because you have the correct amount for the second quest. So on turnin, all of item 1234 would be deleted and you are up the creek trying to turning the second quest. There is never a situation where you would be turning in a quest and need to check for just < and of course we don't care about > :)

runningnak3d committed:
Merge remote-tracking branch 'upstream/3.3.5' into quest6

runningnak3d committed:
Core/Quest: remove another redundant check.

runningnak3d committed:
Core/Quest: Remove redundant operator and handle source items as well.

runningnak3d committed:
Core/Quest: Fix typo == not =, but while fixing typo I realized that if
you have more than needed, the check would fail, hence >=

runningnak3d committed:
Merge remote-tracking branch 'upstream/3.3.5' into quest6

runningnak3d committed:
Merge remote-tracking branch 'upstream/3.3.5' into quest6

Aokromes commented:
@runningnak3d this is completed?

runningnak3d commented:
I listed the tests performed, and the issues it addresses :)

I tried to find any fringe cases where you would want to delete all of an item even though you have other quests in your log that required that item, and I couldn't find any. If I recall, @Shauren even pointed out that these are the things the client checks for ...
